clients/include/host_pinned_vector.hpp

```diff
@@ -0,0 +1,99 @@
+//
+// Copyright 2018-2019 Advanced Micro Devices, Inc.
+//
+#pragma once
+#include "pinned_memory_allocator.hpp"
+#include <memory>
+
+//!
+//! @brief  Pseudo-vector subclass which uses host pinned memory.
+//!
+template <typename T>
+struct host_pinned_vector : std::vector<T, pinned_memory_allocator<T>>
```

Comment:
You do not need to pin memory at allocation time -- it can be pinned after allocation with `mlock()`.

So you don't need custom allocators or separate classes for pinned memory -- it's just an OS call which pins the memory already owned by a data structure.

Reply:
See my previous comment.   What you propose would work but would likely break if hip changes they way they manage memory so is less maintainable.   This simply allocates the memory in the way required, so one could argue is closer to the resource allocation is initialization pattern.

Reply:
`mlock()` is universal in [POSIX](https://pubs.opengroup.org/onlinepubs/9699919799/functions/mlock.html). HIP is unlikely to change this.

If HIP is this fragile, then it needs its own APIs for pinning / unpinning memory, and we use those APIs instead of using POSIX APIs.

Creating new classes or allocators is unnecessarily complicated. Pinning is an operation which is independent from data layout or our own data structures.

Reply:
Also, you can add an extra constructor argument to the existing constructors, if you want to call a different memory allocator.

Something like:
```
 host_vector<T>(HIP_Pinned_Memory, x, ...);
```
and then define an alternative constructor which calls the pinned HIP memory allocator.

There is no need to change the type of the `host_vector` or add extra template parameters. We already have too many types.

If you insist on using a different type name, you can use a derived class:
```
template <typename T>
class host_vector_pinned<T> : public host_vector<T>
{
public:
    template< typename... Ts >
    host_vector_pinned(Ts&&... args) : host_vector<T>( HIP_Pinned_Memory, std::forward<Ts>(args)...) {}
};
```
}